<img src="https://marketing4ecommerce.net/wp-content/uploads/2015/09/logo-iebs.jpg" style="float:right" width="400">

# Introducción a los lenguajes de programación

## Caso Práctico: Estudio de un dataset

### Javier Cózar


# Notas antes de iniciar
Al realizar esta práctica es necesario estar cargando los dataframes cada cierto tiempo debido a que se van modificando línea con línea y en las indicaciones nos marca el original.

De igual manera lo más recomendable es tener el archivo py junto con las hojas de Excel u archivos csv a importar dentro de la misma carpeta para no estar especificando el directorio


# Dataset

En esta libreta vamos a trabajar con un conjunto de datos obtenido de [Kaggle](https://www.kaggle.com/abecklas/fifa-world-cup?select=WorldCupMatches.csv) que contiene los datos de la FIFA de la copa del mundo. Recomendamos acceder a este enlace para descargar los ficheros a trabajar, aunque también disponéis de los archivos ya descargados en la sección de recursos. Utilizaremos estos tres ficheros:

- WorldCupMatches.csv: Datos de los partidos disputados
- WorldCupPlayers.csv: Datos de los jugadores que han jugado en cada partido
- WorldCups.csv: Datos de las copas disputadas y los resultados

# Introducción

Mediante esta libreta vamos a explorar los datos almacenados en estos tres ficheros y a responder preguntas haciendo operaciones y transformaciones con pandas. Además de completar las celdas de código para responder a los ejercicios, cuando así se indique se deberá añadir una celda de tipo _markdown_ explicando con texto enriquecido la información o las conclusiones extraídas.

In [1]:
import pandas as pd

Cargamos el conjunto de datos WorldCupMatches, para esto hay que tener en una carpeta este archivo y la hoja de csv, 
de lo contrario, tendremos que usar r"ruta del archivo"

In [2]:
df_partidos = pd.read_csv("WorldCupMatches.csv")

## 1. Exploración

Usar las funciones `head`, `info` y `describe` para explorar el dataframe `df_partidos`. **Crear una celda de tipo markdown** y comentar brevemente la información almacenada en cada dataframe (columnas y tipo de datos).

_No es necesario hacer una documentación exhaustiva, tan solo comentar aspectos principales como qué tipo de información almacena el DataFrame, cuántas filas y columnas tiene, y comentar las columnas que se consideren más relevantes._

In [3]:
df_partidos.describe(include="all")

Year              Datetime        Stage         Stadium  \
count    852.000000                   852          852             852   
unique          NaN                   602           23             181   
top             NaN  27 May 1934 - 16:30   Round of 16  Estadio Azteca   
freq            NaN                     8           72              19   
mean    1985.089202                   NaN          NaN             NaN   
std       22.448825                   NaN          NaN             NaN   
min     1930.000000                   NaN          NaN             NaN   
25%     1970.000000                   NaN          NaN             NaN   
50%     1990.000000                   NaN          NaN             NaN   
75%     2002.000000                   NaN          NaN             NaN   
max     2014.000000                   NaN          NaN             NaN   

                City Home Team Name  Home Team Goals  Away Team Goals  \
count            852            852       852.000000       852.000000   
unique           151             78              NaN              NaN   
top     Mexico City          Brazil              NaN              NaN   
freq              23             82              NaN              NaN   
mean             NaN            NaN         1.811033         1.022300   
std              NaN            NaN         1.610255         1.087573   
min              NaN            NaN         0.000000         0.000000   
25%              NaN            NaN         1.000000         0.000000   
50%              NaN            NaN         2.000000         1.000000   
75%              NaN            NaN         3.000000         2.000000   
max              NaN            NaN        10.000000         7.000000   

       Away Team Name Win conditions     Attendance  Half-time Home Goals  \
count             852            852     850.000000            852.000000   
unique             83             43            NaN                   NaN   
top            Mexico                           NaN                   NaN   
freq               38            787            NaN                   NaN   
mean              NaN            NaN   45164.800000              0.708920   
std               NaN            NaN   23485.249247              0.937414   
min               NaN            NaN    2000.000000              0.000000   
25%               NaN            NaN   30000.000000              0.000000   
50%               NaN            NaN   41579.500000              0.000000   
75%               NaN            NaN   61374.500000              1.000000   
max               NaN            NaN  173850.000000              6.000000   

        Half-time Away Goals                Referee  \
count             852.000000                    852   
unique                   NaN                    366   
top                      NaN  Ravshan IRMATOV (UZB)   
freq                     NaN                     10   
mean                0.428404                    NaN   
std                 0.691252                    NaN   
min                 0.000000                    NaN   
25%                 0.000000                    NaN   
50%                 0.000000                    NaN   
75%                 1.000000                    NaN   
max                 5.000000                    NaN   

                             Assistant 1               Assistant 2  \
count                                852                       852   
unique                               387                       408   
top     GONZALEZ ARCHUNDIA Alfonso (MEX)  KOCHKAROV Bakhadyr (KGZ)   
freq                                   7                        10   
mean                                 NaN                       NaN   
std                                  NaN                       NaN   
min                                  NaN                       NaN   
25%                                  NaN                       NaN   
50%                              

head() - Al explorar los datos con head() nos muetras principalmente el encabezado de la tabla, es decir el nombre de las columnas, donde podemos observar manualmente haciendo scroll a la nbarra cuantas hay y unos datos de ellas, estos datos o filas que aprecen por defecto son 5, sin embargo, podemos especificar cuantas filas queremos ver poniendo un numero dentro del parentesis
ejemplo: df_partidos.head(10) e iniciara del 0 al 9 

info() - se puede observar que hay un total de 20 columnas donde algunas son de tipo float64 u object es decir de tipo flotante o string correspondientemente 
 
 0   Year                  852 non-null    float64
 1   Datetime              852 non-null    object 
 2   Stage                 852 non-null    object 
 3   Stadium               852 non-null    object 
 4   City                  852 non-null    object 
 5   Home Team Name        852 non-null    object 
 6   Home Team Goals       852 non-null    float64
 7   Away Team Goals       852 non-null    float64
 8   Away Team Name        852 non-null    object 
 9   Win conditions        852 non-null    object 
 10  Attendance            850 non-null    float64
 11  Half-time Home Goals  852 non-null    float64
 12  Half-time Away Goals  852 non-null    float64
 13  Referee               852 non-null    object 
 14  Assistant 1           852 non-null    object 
 15  Assistant 2           852 non-null    object 
 16  RoundID               852 non-null    float64
 17  MatchID               852 non-null    float64
 18  Home Team Initials    852 non-null    object 
 19  Away Team Initials

describe(include="all") -Lo destacable aqui es que podemos ver que hay 852 datos en cada columna, dentro de los datos mas repetidos estan 
Datetime 27 May 1934 - 16:30 con 8
Stage Round of 16 con 72
Stadium Estadio Azteca con 19
City Mexico con 23 
Home Team Name Brazil con 82
Away Team Name Mexico  con 38
Referee Ravshan IRMATOV (UZB) con 10 
Assistant 1 BERANEK Alois (AUT) con 7 
Assistant 2 KOCHKAROV Bakhadyr (KGZ) con 10 	
Home Team Initials BRA	con 82
Away Team Initials Mex con 38

## 2. Limpieza de datos

Una de las cosas que llama la atención es la alta presencia de valores perdidos en el DataFrame `df_partidos`. Cuando trabajamos con datos obtenidos del mundo real siempre nos toparemos con problemas relacionados con la medición, captura o almacenamiento de dicha información.

Localiza las filas con valores perdido. Analizar y **documentar en una nueva celda de tipo markdown** a qué se deben estos valores perdidos. Finalmente usar la función [dropna](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dropna.html) sobre el DataFrame `df_partidos` para eliminar los valores perdidos del DataFrame, y almacena el resultado en `df_partidos` de nuevo.

In [4]:
df_partidos.isnull().sum()

Year                    0
Datetime                0
Stage                   0
Stadium                 0
City                    0
Home Team Name          0
Home Team Goals         0
Away Team Goals         0
Away Team Name          0
Win conditions          0
Attendance              2
Half-time Home Goals    0
Half-time Away Goals    0
Referee                 0
Assistant 1             0
Assistant 2             0
RoundID                 0
MatchID                 0
Home Team Initials      0
Away Team Initials      0
dtype: int64

In [5]:
df_partidos[df_partidos.Attendance.isnull()]

Year              Datetime        Stage            Stadium  \
823  2014  30 Jun 2014 - 17:00   Round of 16  Estadio Beira-Rio   
841  2014  30 Jun 2014 - 17:00   Round of 16  Estadio Beira-Rio   

              City Home Team Name  Home Team Goals  Away Team Goals  \
823  Porto Alegre         Germany                2                1   
841  Porto Alegre         Germany                2                1   

    Away Team Name                 Win conditions  Attendance  \
823        Algeria  Germany win after extra time          NaN   
841        Algeria  Germany win after extra time          NaN   

     Half-time Home Goals  Half-time Away Goals             Referee  \
823                     0                     0  RICCI Sandro (BRA)   
841                     0                     0  RICCI Sandro (BRA)   

                   Assistant 1              Assistant 2  RoundID    MatchID  \
823  DE CARVALHO Emerson (BRA)  VAN GASSE Marcelo (BRA)   255951  300186460   
841  DE CARVALHO Emerson (BRA)  VAN GASSE Marcelo (BRA)   255951  300186460   

    Home Team Initials Away Team Initials  
823                GER                ALG  
841                GER                ALG

Al ejecutar isnull().sum() podemos observar que hay dos valores perdidos en la columna attendance y para poder visualizar estas columnas se usa df_partidos[df_partidos.Attendance.isnull()]. Como no hay valor alguno marca NaN, esto se puede veridicar en el excell, donde tendremos que buscar la celda agregando dos , es decir 823+2 y 841+2, est debido al encabezado y a que la numeracion en python corre desde 0 y no 1 como excell  

Nota: Antes de correr isnull().sum() es recomendable editar y eliminar en un block de notas los valores que estan completamente vacios, los que solo tienen comas para ver el numero real de valores perdiddos y no las celdas completas de excell


In [6]:
df_partidos_corregidos=df_partidos.dropna(axis=0)
df_partidos_corregidos

Year              Datetime                     Stage  \
0    1930  13 Jul 1930 - 15:00                    Group 1   
1    1930  13 Jul 1930 - 15:00                    Group 4   
2    1930  14 Jul 1930 - 12:45                    Group 2   
3    1930  14 Jul 1930 - 14:50                    Group 3   
4    1930  15 Jul 1930 - 16:00                    Group 1   
..    ...                   ...                       ...   
847  2014  05 Jul 2014 - 17:00             Quarter-finals   
848  2014  08 Jul 2014 - 17:00                Semi-finals   
849  2014  09 Jul 2014 - 17:00                Semi-finals   
850  2014  12 Jul 2014 - 17:00   Play-off for third place   
851  2014  13 Jul 2014 - 16:00                      Final   

                 Stadium             City Home Team Name  Home Team Goals  \
0                Pocitos      Montevideo          France                4   
1         Parque Central      Montevideo             USA                3   
2         Parque Central      Montevideo      Yugoslavia                2   
3                Pocitos      Montevideo         Romania                3   
4         Parque Central      Montevideo       Argentina                1   
..                   ...              ...            ...              ...   
847     Arena Fonte Nova        Salvador     Netherlands                0   
848     Estadio Mineirao  Belo Horizonte          Brazil                1   
849   Arena de Sao Paulo       Sao Paulo     Netherlands                0   
850     Estadio Nacional        Brasilia          Brazil                0   
851  Estadio do Maracana  Rio De Janeiro         Germany                1   

     Away Team Goals Away Team Name                         Win conditions  \
0                  1         Mexico                                          
1                  0        Belgium                                          
2                  1         Brazil                                          
3                  1           Peru                                          
4                  0         France                                          
..               ...            ...                                    ...   
847                0     Costa Rica  Netherlands win on penalties (4 - 3)    
848                7        Germany                                          
849                0      Argentina    Argentina win on penalties (2 - 4)    
850                3    Netherlands                                          
851                0      Argentina          Germany win after extra time    

     Attendance  Half-time Home Goals  Half-time Away Goals  \
0        4444.0                     3                     0   
1       18346.0                     2                     0   
2       24059.0                     2                     0   
3        2549.0                     1                     0   
4       23409.0                     0                     0   
..          ...                   ...                   ...   
847     51179.0                     0                     0   
848     58141.0                     0                     5   
849     63267.0                     0                     0   
850     68034.0                     0                     2   
851     74738.0                     0                     0   

                    Referee                  Assistant 1  \
0    LOMBARDI Domingo (URU)        CRISTOPHE Henry (BEL)   
1         MACIAS Jose (ARG)     MATEUCCI Francisco (URU)   
2       TEJADA Anibal (URU)      VALLARINO Ricardo (URU)   
3     WARNKEN Alberto (CHI)          LANGENUS Jean (BEL)   
4       REGO Gilberto (BRA)         SAUCEDO Ulises (BOL)   
..                      ...                          ...   
847   Ravshan IRMATOV (UZB)  RASULOV Abduxamidullo (UZB)   
848   RODRIGUEZ Marco (MEX)      TORRENTERA Marvin (MEX)   
849      C�neyt �AKIR (TUR)         DURAN Bahattin (TUR)   
850   HAIMOUDI Djamel (ALG)         ACHIK Redouane (M

Como se puede ver en la informacion, se han eliminado dos filas

## 3. Cargar datos

Al igual que hemos cargado los datos de los partidos en `df_partidos`, crear las variables `df_jugadores` y `df_copas` que contengan los dataframes correspondientes a la lectura de los csv `WorldCupPlayers` y `WorldCups` (usar `pd.read_csv)`.


In [7]:
df_jugadores = pd.read_csv("WorldCupPlayers.csv")
df_copas = pd.read_csv("WorldCups.csv")

## 4. Rango temporal

El dataframe `df_copas` contiene datos de todos los mundiales disputados. ¿Cuál es el año del mundial más antiguo disputado? ¿Y el año del mundial más reciente?

**Pista:** recuerda que podemos usar funciones de agregación (`count` para contar el número de casos, `mean` para calcular el valor medio, etc.) directamente sobre columnas o Series de pandas. Por ejemplo, el siguiente código nos muestra el mayor número de goles marcado en un mundial.

```python
df_copas["GoalsScored"].max()
```

**Opcional:** Una vez localizados los años de interés (más antiguo y más reciente) visualizar las filas completas correspondientes a cada año usando la función `.loc`.

In [8]:
df_copas["Year"].min()


1930

In [9]:
df_copas["Year"].max()

2014

In [10]:
# df_copas.describe(include="all")

In [11]:
df_copas

Year       Country      Winner      Runners-Up        Third  \
0   1930       Uruguay     Uruguay       Argentina          USA   
1   1934         Italy       Italy  Czechoslovakia      Germany   
2   1938        France       Italy         Hungary       Brazil   
3   1950        Brazil     Uruguay          Brazil       Sweden   
4   1954   Switzerland  Germany FR         Hungary      Austria   
5   1958        Sweden      Brazil          Sweden       France   
6   1962         Chile      Brazil  Czechoslovakia        Chile   
7   1966       England     England      Germany FR     Portugal   
8   1970        Mexico      Brazil           Italy   Germany FR   
9   1974       Germany  Germany FR     Netherlands       Poland   
10  1978     Argentina   Argentina     Netherlands       Brazil   
11  1982         Spain       Italy      Germany FR       Poland   
12  1986        Mexico   Argentina      Germany FR       France   
13  1990         Italy  Germany FR       Argentina        Italy   
14  1994           USA      Brazil           Italy       Sweden   
15  1998        France      France          Brazil      Croatia   
16  2002   Korea/Japan      Brazil         Germany       Turkey   
17  2006       Germany       Italy          France      Germany   
18  2010  South Africa       Spain     Netherlands      Germany   
19  2014        Brazil     Germany       Argentina  Netherlands   

            Fourth  GoalsScored  QualifiedTeams  MatchesPlayed Attendance  
0       Yugoslavia           70              13             18    590.549  
1          Austria           70              16             17    363.000  
2           Sweden           84              15             18    375.700  
3            Spain           88              13             22  1.045.246  
4          Uruguay          140              16             26    768.607  
5       Germany FR          126              16             35    819.810  
6       Yugoslavia           89              16             32    893.172  
7     Soviet Union           89              16             32  1.563.135  
8          Uruguay           95              16             32  1.603.975  
9           Brazil           97              16             38  1.865.753  
10           Italy          102              16             38  1.545.791  
11          France          146              24             52  2.109.723  
12         Belgium          132              24             52  2.394.031  
13         England          115              24             52  2.516.215  
14        Bulgaria          141              24             52  3.587.538  
15     Netherlands          171              32             64  2.785.100  
16  Korea Republic          161              32             64  2.705.197  
17        Portugal          147              32             64  3.359.439  
18         Uruguay          145              32             64  3.178.856  
19          Brazil          171              32             64  3.386.810

In [12]:
df_copas.iloc[[0,-1]]
#df_copas.set_index("Year",inplace=True)
#df_copas.loc[[1930,2014]]

Year  Country   Winner Runners-Up        Third      Fourth  GoalsScored  \
0   1930  Uruguay  Uruguay  Argentina          USA  Yugoslavia           70   
19  2014   Brazil  Germany  Argentina  Netherlands      Brazil          171   

    QualifiedTeams  MatchesPlayed Attendance  
0               13             18    590.549  
19              32             64  3.386.810

El año del mundial mas antiguo es de 1930 y el mas actual registrado es de 2014,
si bien se puede usar df_copas["Year"].min() y df_copas["Year"].max() para obtener el año mas antiguo y mas reciente de igual manera se puede usar describe y describe nos da mayor informacion ademas de estas dos juntas.

En lugar de usar la funcion loc, en esta ocasion es mas facil usar iloc especificando los indices para imprimir el mas antiguo y el mas actual (0,-1) porque la tabla original registra los mundiales consecutivamente.

si se quiere usar loc es necesario especificar en que columna buscaremos el dato 
df_copas.set_index("Year",inplace=True)
y usar la funcion loc para buscar las filas corresponidnetes a esos datos 
df_copas.loc[[1930,2014]]

---
# Parte opcional

Los ejercicios anteriores son obligatorios y puntuan hasta un máximo de 80 puntos sobre 100.

Los siguientes ejercicios trabajan con los dataframes cargados para extraer información concreta a partir de los datos. 
Estos ejercicios son opcionales, pero es recomendable intentar resolverlos para adquirir habilidades.

---

## 5. Asistencia de público

La columna `Attendance` indica la asistencia de público en cada mundial. Pandas ha cargado esta columna como tipo string, ya que los valores indicados contienen puntos como separador de millares. En primer lugar vamos a transformar esta columna para que no contenga el caracter `.`. Posteriormente transformaremos esta columna a formato numérico usando la función [pd.to_numeric](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.to_numeric.html).

**Nota**: la función `pd.read_csv` también permite hacer este proceso de una forma mucho más transparente, usando el argumento `thousands`. Ver la documentación [aquí](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html).

In [13]:
#df_copas["Attendance"] = pd.to_numeric(df_copas["Attendance"].str.replace(".", ""))
df_copas = pd.read_csv("WorldCups.csv", thousands=".")

In [14]:
type(df_copas.Attendance[1])

numpy.int64

Teniendo en cuenta todos los datos de `df_copas`, ¿Cuál fué la mayor asistencia de un mundial? ¿Cuál es la asistencia media?

In [15]:
df_copas.sort_values("Attendance",ascending=False)

Year       Country      Winner      Runners-Up        Third  \
14  1994           USA      Brazil           Italy       Sweden   
19  2014        Brazil     Germany       Argentina  Netherlands   
17  2006       Germany       Italy          France      Germany   
18  2010  South Africa       Spain     Netherlands      Germany   
15  1998        France      France          Brazil      Croatia   
16  2002   Korea/Japan      Brazil         Germany       Turkey   
13  1990         Italy  Germany FR       Argentina        Italy   
12  1986        Mexico   Argentina      Germany FR       France   
11  1982         Spain       Italy      Germany FR       Poland   
9   1974       Germany  Germany FR     Netherlands       Poland   
8   1970        Mexico      Brazil           Italy   Germany FR   
7   1966       England     England      Germany FR     Portugal   
10  1978     Argentina   Argentina     Netherlands       Brazil   
3   1950        Brazil     Uruguay          Brazil       Sweden   
6   1962         Chile      Brazil  Czechoslovakia        Chile   
5   1958        Sweden      Brazil          Sweden       France   
4   1954   Switzerland  Germany FR         Hungary      Austria   
0   1930       Uruguay     Uruguay       Argentina          USA   
2   1938        France       Italy         Hungary       Brazil   
1   1934         Italy       Italy  Czechoslovakia      Germany   

            Fourth  GoalsScored  QualifiedTeams  MatchesPlayed  Attendance  
14        Bulgaria          141              24             52     3587538  
19          Brazil          171              32             64     3386810  
17        Portugal          147              32             64     3359439  
18         Uruguay          145              32             64     3178856  
15     Netherlands          171              32             64     2785100  
16  Korea Republic          161              32             64     2705197  
13         England          115              24             52     2516215  
12         Belgium          132              24             52     2394031  
11          France          146              24             52     2109723  
9           Brazil           97              16             38     1865753  
8          Uruguay           95              16             32     1603975  
7     Soviet Union           89              16             32     1563135  
10           Italy          102              16             38     1545791  
3            Spain           88              13             22     1045246  
6       Yugoslavia           89              16             32      893172  
5       Germany FR          126              16             35      819810  
4          Uruguay          140              16             26      768607  
0       Yugoslavia           70              13             18      590549  
2           Sweden           84              15             18      375700  
1          Austria           70              16             17      363000

In [16]:
 pd.set_option('float_format', '{:.2f}'.format)

In [17]:
df_copas.describe(include="all")

Year  Country  Winner Runners-Up    Third   Fourth  GoalsScored  \
count    20.00       20      20         20       20       20        20.00   
unique     nan       15       9         10       14       16          nan   
top        nan  Germany  Brazil  Argentina  Germany  Uruguay          nan   
freq       nan        2       5          3        3        3          nan   
mean   1974.80      NaN     NaN        NaN      NaN      NaN       118.95   
std      25.58      NaN     NaN        NaN      NaN      NaN        32.97   
min    1930.00      NaN     NaN        NaN      NaN      NaN        70.00   
25%    1957.00      NaN     NaN        NaN      NaN      NaN        89.00   
50%    1976.00      NaN     NaN        NaN      NaN      NaN       120.50   
75%    1995.00      NaN     NaN        NaN      NaN      NaN       145.25   
max    2014.00      NaN     NaN        NaN      NaN      NaN       171.00   

        QualifiedTeams  MatchesPlayed  Attendance  
count            20.00          20.00       20.00  
unique             nan            nan         nan  
top                nan            nan         nan  
freq               nan            nan         nan  
mean             21.25          41.80  1872882.35  
std               7.27          17.22  1071842.15  
min              13.00          17.00   363000.00  
25%              16.00          30.50   874831.50  
50%              16.00          38.00  1734864.00  
75%              26.00          55.00  2725172.75  
max              32.00          64.00  3587538.00

La mayor asistencia ha sido de 3587538 en el munidal de 1994,USA.

La asistencia media ha sido de 1872882.35

## 6. Asistencia de público en función del país

Sabemos que, dependiendo del país donde se celebre la copa del mundo, ésta tiene un mayor o menor impacto o relevancia. Vamos a calcular la media de asistencia agrupando por país. Finalmente, ordenar los resultados por orden decreciente de este número medio de asistentes.

**Pista 1:** Como se muestra en el siguiente ejemplo, recuerda que podemos combinar la función [groupby](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html) con otras funciones de agregación como `mean`.

**Pista 2:** Recuerda que podemos usar la función [sort_values](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sort_values.html) para ordenar un DataFrame por una de sus columnas. Por ejemplo, el siguiente código muestra los mundiales ordenados de forma decreciente por el número de goles marcados:

```python
df_copas.sort_values("GoalsScored", ascending=False)
```

In [18]:
df_personas = pd.DataFrame(
    [
        ("Juan", "male", 56),
        ("Laura", "female", 23),
        ("José", "male", 46),
        ("Rosa", "female", 26)
    ], columns=["name", "gender", "age"])

# edad media de todas las personas
edad_media = df_personas["age"].mean()
# edad media agrupando por género
edad_media_por_genero = df_personas.groupby("gender").mean()[["age"]]

print(f"La edad media de las {len(df_personas)} es de {edad_media} años.")
print(f"La edad media agrupada por género es:")
edad_media_por_genero

La edad media de las 4 es de 37.75 años.
La edad media agrupada por género es:


age
gender      
female 24.50
male   51.00

In [19]:
df_copas = pd.read_csv("WorldCups.csv")
df_copas = pd.read_csv("WorldCups.csv", thousands=".")
edad_media = df_copas["Attendance"].mean()
df_copas = df_copas.groupby("Country").mean()[["Attendance"]]
print(f"La asistencia media registrada en los {len(df_copas)} paises diferentes que se ha jugado la copa mundial, es de {edad_media} personas.")
print(f"La asistencia media agrupada por pais es:")
df_copas.sort_values("Attendance",ascending=False)

La asistencia media registrada en los 15 paises diferentes que se ha jugado la copa mundial, es de 1872882.35 personas.
La asistencia media agrupada por pais es:


Attendance
Country                 
USA           3587538.00
South Africa  3178856.00
Korea/Japan   2705197.00
Germany       2612596.00
Brazil        2216028.00
Spain         2109723.00
Mexico        1999003.00
France        1580400.00
England       1563135.00
Argentina     1545791.00
Italy         1439607.50
Chile          893172.00
Sweden         819810.00
Switzerland    768607.00
Uruguay        590549.00

## 7. País con más victorias

Vamos a analizar el número de veces que un país ha ganado la copa del mundo. Transformar el DataFrame `df_copas` para que muestre los diferentes países que han ganado al menos una copa del mundo, y ordena el DataFrame por dicho número de forma descendente usando la función [sort_values](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sort_values.html).

¿Ves algo extraño o digno de comentar? Pista: analiza Alemania. **Hazlo en una nueva celda de tipo markdown.**

In [20]:
df_copas = pd.read_csv("WorldCups.csv")
df_copas.describe(include="all")

Year  Country  Winner Runners-Up    Third   Fourth  GoalsScored  \
count    20.00       20      20         20       20       20        20.00   
unique     nan       15       9         10       14       16          nan   
top        nan  Germany  Brazil  Argentina  Germany  Uruguay          nan   
freq       nan        2       5          3        3        3          nan   
mean   1974.80      NaN     NaN        NaN      NaN      NaN       118.95   
std      25.58      NaN     NaN        NaN      NaN      NaN        32.97   
min    1930.00      NaN     NaN        NaN      NaN      NaN        70.00   
25%    1957.00      NaN     NaN        NaN      NaN      NaN        89.00   
50%    1976.00      NaN     NaN        NaN      NaN      NaN       120.50   
75%    1995.00      NaN     NaN        NaN      NaN      NaN       145.25   
max    2014.00      NaN     NaN        NaN      NaN      NaN       171.00   

        QualifiedTeams  MatchesPlayed Attendance  
count            20.00          20.00         20  
unique             nan            nan         20  
top                nan            nan  3.178.856  
freq               nan            nan          1  
mean             21.25          41.80        NaN  
std               7.27          17.22        NaN  
min              13.00          17.00        NaN  
25%              16.00          30.50        NaN  
50%              16.00          38.00        NaN  
75%              26.00          55.00        NaN  
max              32.00          64.00        NaN

In [21]:
df_copas["Winner"].value_counts(ascending=True)

Germany       1
England       1
Spain         1
France        1
Argentina     2
Uruguay       2
Germany FR    3
Italy         4
Brazil        5
Name: Winner, dtype: int64

El pais con que mayor veces ha ganado la copa es Brazil, sin contar alemania occidental y Alemania como distintos, estos tienen la mismas copas que italia, tan solo una menos que Brazil

## Creación de un nuevo DataFrame

A continuación vamos a trabajar con el DataFrame `df_partidos`. Vamos a centrarnos en la información a nivel de selección, siendo indiferente si el equipo juega en casa o fuera. En este sentido, la información disponible no es la mejor para trabajar cómodamente. Por ello, es habitual realizar transformaciones a los datos en función de las preguntas que le queramos hacer. A continuación crearemos un nuevo DataFrame llamado `df_paises`que contendrá todos los paises involucrados en la copa del mundo y el año en el que participaron. Es decir, un país que haya jugado varias copas del mundo aparecerá varias veces:


||Year|Country|
|---|---|---|
|0|1930|France|
|1|1930|USA|
|...|...|..|
|37|1938|France|
|...|...|..|

**Nota:** este problema se puede afrontar como mínimo de dos maneras diferentes

1. Generando dos dataframes, uno donde la columna `Country` se corresponde con `Home Team Name`, y otro donde la columna `Country` se corresponde con `Away Team Name`. Finalmente, combinar ambos DataFrames en uno solo.
2. Usar la función [melt](https://pandas.pydata.org/docs/reference/api/pandas.melt.html) que nos permite convertir N columnas en un par de columnas llamadas `variable` y `value`, donde la primera es el nombre de una de las columnas y `value` el valor que tomó dicha variable.


In [22]:
#alternativa 1
df_paises = (
    pd.concat([
        df_partidos[["Year", "Home Team Name"]].rename(columns={"Home Team Name": "Country"}),
        df_partidos[["Year", "Away Team Name"]].rename(columns={"Away Team Name": "Country"})
    ])
)
#alternativa 2
df_paises = (
    df_partidos
    .melt(id_vars=["Year"], value_vars=["Home Team Name", "Away Team Name"])
    .rename(columns={"value": "Country"})
    [["Year", "Country"]]
)

In [23]:
df_paises

Year      Country
0     1930       France
1     1930          USA
2     1930   Yugoslavia
3     1930      Romania
4     1930    Argentina
...    ...          ...
1699  2014   Costa Rica
1700  2014      Germany
1701  2014    Argentina
1702  2014  Netherlands
1703  2014    Argentina

[1704 rows x 2 columns]

## 8. Número de países

Vamos a descubrir cuántos países han participado, al menos una vez, en una copa del mundo. Para ello vamos a trabajar con el DataFrame construido anteriormente y a usar la función de agregación [nunique](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.nunique.html).

In [24]:
df_paises.nunique()

Year       20
Country    83
dtype: int64

Para saber unicamente cuantos paises han participado se ejecuta la siguinete linea de codigo

In [25]:
#df_paises["Country"].nunique()

Para obtener los nombres de los paises que han participado 

In [26]:
#df_paises["Country"].unique()

Han participado 83 paises en los 20 mundiaes registrados.

La lista de paises que han participado al menos en un mundial son:

'France', 'USA', 'Yugoslavia', 'Romania', 'Argentina', 'Chile','Uruguay', 'Brazil', 'Paraguay', 'Austria', 'Hungary','Switzerland', 'Sweden', 'Germany', 'Spain', 'Italy','Czechoslovakia', 'Cuba', 'England', 'Germany FR', 'Turkey','Soviet Union', 'Northern Ireland', 'Mexico', 'Wales', 'Portugal','Korea DPR', 'Peru', 'Belgium', 'Bulgaria', 'German DR', 'Zaire','Poland', 'Australia', 'Scotland', 'Netherlands', 'Haiti','Tunisia', 'Algeria', 'Honduras', 'Canada', 'Morocco','Korea Republic', 'Iraq', 'Denmark', 'rn">United Arab Emirates','Costa Rica', 'Cameroon', 'rn">Republic of Ireland', 'Colombia','Norway', 'Nigeria', 'Saudi Arabia', 'Bolivia', 'Russia', 'Greece','Jamaica', 'South Africa', 'Japan', 'Croatia', 'China PR','Senegal', 'Slovenia', 'Ecuador', 'rn">Trinidad and Tobago','rn">Serbia and Montenegro', 'Angola', 'Czech Republic', 'Togo','Iran', "C�te d'Ivoire", 'Ghana', 'Ukraine', 'Serbia','New Zealand', 'Slovakia', 'IR Iran', 'rn">Bosnia and Herzegovina','Egypt', 'Dutch East Indies', 'Israel', 'El Salvador', 'Kuwait'

## 9. Número de partidos por país

Ahora vamos a descubrir cuántos partidos ha jugado cada país. Operar con el DataFrame para que muestre un país por fila y una columna que indique el número de partidos que ha jugado. Ordenar el DataFrame por dicha columna de forma descendente.

In [27]:
pd.set_option("max_rows", 83)
df_paises["Country"].value_counts(ascending=False)

Brazil                        108
Italy                          83
Argentina                      81
Germany FR                     62
England                        62
France                         61
Spain                          59
Netherlands                    54
Mexico                         54
Uruguay                        52
Germany                        48
Sweden                         46
Belgium                        43
Yugoslavia                     37
USA                            34
Switzerland                    34
Chile                          34
Hungary                        32
Poland                         31
Korea Republic                 31
Soviet Union                   31
Czechoslovakia                 30
Austria                        29
Paraguay                       27
Bulgaria                       26
Portugal                       26
Scotland                       23
Cameroon                       23
Romania                        21
Colombia      

El pais que mas veces ha jugado partidos es Brazil con 108, mientras que el que menos ha jugado es Dutch East Indies  con 1, cabe mencionar que ya no existe, ahora es indonecia.

Si omitimos este pais, hay varios que solo han jugado 3 partidos en los que se encuentran Canada,Israel,Cuba,Haiti,Iraq, entre otros.


## 10. Partidos jugados por España

Vamos a trabajar a continuación con el DataFrame `df_partidos`. Seleccionar aquellos partidos (filas del dataframe) en los que España ha jugado como `Home Team`, es decir, donde la columna `Home Team` tiene el valor `España`. Recuerda que para la operación de filtrado hay que usar la función [.loc](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.loc.html).


A continuación, haz una nueva selección en los que España ha jugado como `Away Team`, es decir, donde la columna `Away Team` tiene el valor `España`.

**Opcional:** ¿Podrías hacer una sola selección donde España haya jugado como `Home Team` o como `Away Team`?. Pista: consulta la documentación del operador de [disyunción](https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#boolean-indexing).

In [28]:
df_partidos.set_index("Away Team Name",inplace=True)
df_partidos.loc[["Spain"]]

Year              Datetime           Stage  \
Away Team Name                                               
Spain           1934  31 May 1934 - 16:30   Quarter-finals   
Spain           1934  01 Jun 1934 - 16:30   Quarter-finals   
Spain           1950  09 Jul 1950 - 15:00          Group 6   
Spain           1950  13 Jul 1950 - 15:00          Group 6   
Spain           1950  16 Jul 1950 - 15:00          Group 6   
Spain           1962  31 May 1962 - 15:00          Group 3   
Spain           1962  06 Jun 1962 - 15:00          Group 3   
Spain           1966  13 Jul 1966 - 19:30          Group 2   
Spain           1966  20 Jul 1966 - 19:30          Group 2   
Spain           1978  03 Jun 1978 - 13:45          Group 3   
Spain           1978  07 Jun 1978 - 13:45          Group 3   
Spain           1982  25 Jun 1982 - 21:00          Group 5   
Spain           1982  02 Jul 1982 - 21:00          Group 2   
Spain           1986  07 Jun 1986 - 12:00          Group D   
Spain           1986  12 Jun 1986 - 12:00          Group D   
Spain           1986  18 Jun 1986 - 16:00      Round of 16   
Spain           1990  13 Jun 1990 - 17:00          Group E   
Spain           1990  17 Jun 1990 - 21:00          Group E   
Spain           1990  21 Jun 1990 - 17:00          Group E   
Spain           1994  21 Jun 1994 - 16:00          Group C   
Spain           1994  27 Jun 1994 - 16:00          Group C   
Spain           1994  09 Jul 1994 - 12:00   Quarter-finals   
Spain           2002  12 Jun 2002 - 20:30          Group B   
Spain           2006  23 Jun 2006 - 16:00          Group H   
Spain           2010  25 Jun 2010 - 20:30          Group H   
Spain           2010  03 Jul 2010 - 20:30   Quarter-finals   
Spain           2010  07 Jul 2010 - 20:30      Semi-finals   
Spain           2010  11 Jul 2010 - 20:30            Final   
Spain           2014  23 Jun 2014 - 13:00          Group B   

                                                  Stadium               City  \
Away Team Name                                                                 
Spain                                      Giovanni Berta          Florence    
Spain                                      Giovanni Berta          Florence    
Spain                                            Pacaembu         Sao Paulo    
Spain           Maracan� - Est�dio Jornalista M�rio Filho    Rio De Janeiro    
Spain                                            Pacaembu         Sao Paulo    
Spain                                   Estadio Sausalito      Vina Del Mar    
Spain                                   Estadio Sausalito      Vina Del Mar    
Spain                                          Villa Park        Birmingham    
Spain                                          Villa Park        Birmingham    
Spain                                     Jose Amalfitani      Buenos Aires    
Spain                          Estadio Jos� Mar�a Minella     Mar Del Plata    
Spain                                       Luis Casanova          Valencia    
Spain                                   Santiago Bernabeu            Madrid    
Spain                                       Tres de Marzo       Guadalajara    
Spain                                         Tecnologico         Monterrey    
Spain                                 Estadio Corregidora         Queretaro    
Spain                                         Dacia Arena             Udine    
Spain                                         Dacia Arena             Udine    
Spain                              Marc Antonio Bentegodi            Verona    
Spain                                       Soldier Field           Chicago    
Spain                                       Soldier Field           Chicago    
Spain                                     Foxboro Stadium            Boston    
Spain                           Daejeon World Cup Stadium           Daejeon    
Spain                                Fritz-Walter-Stadion    Kaiserslautern    
Spain        

In [29]:
df_partidos.set_index("Home Team Name",inplace=True)
df_partidos.loc[["Spain"]]

Year               Datetime              Stage  \
Home Team Name                                                   
Spain           1934   27 May 1934 - 16:30   Preliminary round   
Spain           1950   25 Jun 1950 - 15:00             Group 2   
Spain           1950   29 Jun 1950 - 15:00             Group 2   
Spain           1950   02 Jul 1950 - 15:00             Group 2   
Spain           1962   03 Jun 1962 - 15:00             Group 3   
Spain           1966   15 Jul 1966 - 19:30             Group 2   
Spain           1978   11 Jun 1978 - 13:45             Group 3   
Spain           1982   16 Jun 1982 - 21:00             Group 5   
Spain           1982   20 Jun 1982 - 21:00             Group 5   
Spain           1982   05 Jul 1982 - 21:00             Group 2   
Spain           1986   01 Jun 1986 - 12:00             Group D   
Spain           1986   22 Jun 1986 - 16:00      Quarter-finals   
Spain           1990   26 Jun 1990 - 17:00         Round of 16   
Spain           1994   17 Jun 1994 - 19:30             Group C   
Spain           1994   02 Jul 1994 - 16:30         Round of 16   
Spain           1998   13 Jun 1998 - 14:30             Group D   
Spain           1998   19 Jun 1998 - 21:00             Group D   
Spain           1998   24 Jun 1998 - 21:00             Group D   
Spain           2002   02 Jun 2002 - 20:30             Group B   
Spain           2002  07 June 2002 - 18:00             Group B   
Spain           2002   16 Jun 2002 - 20:30         Round of 16   
Spain           2002   22 Jun 2002 - 15:30      Quarter-finals   
Spain           2006   14 Jun 2006 - 15:00             Group H   
Spain           2006   19 Jun 2006 - 21:00             Group H   
Spain           2006   27 Jun 2006 - 21:00         Round of 16   
Spain           2010   16 Jun 2010 - 16:00             Group H   
Spain           2010   21 Jun 2010 - 20:30             Group H   
Spain           2010   29 Jun 2010 - 20:30         Round of 16   
Spain           2014   13 Jun 2014 - 16:00             Group B   
Spain           2014   18 Jun 2014 - 16:00             Group B   

                                                  Stadium             City  \
Home Team Name                                                               
Spain                                      Luigi Ferraris           Genoa    
Spain                                    Durival de Brito        Curitiba    
Spain           Maracan� - Est�dio Jornalista M�rio Filho  Rio De Janeiro    
Spain           Maracan� - Est�dio Jornalista M�rio Filho  Rio De Janeiro    
Spain                                   Estadio Sausalito    Vina Del Mar    
Spain                                        Hillsborough       Sheffield    
Spain                                     Jose Amalfitani    Buenos Aires    
Spain                                       Luis Casanova        Valencia    
Spain                                       Luis Casanova        Valencia    
Spain                                   Santiago Bernabeu          Madrid    
Spain                                             Jalisco     Guadalajara    
Spain                                          Cuauhtemoc          Puebla    
Spain                              Marc Antonio Bentegodi          Verona    
Spain                                         Cotton Bowl          Dallas    
Spain                                         RFK Stadium   Washington Dc    
Spain                                        La Beaujoire          Nantes    
Spain                             Stade Geoffroy Guichard   Saint-Etienne    
Spain                                Stade Felix Bollaert            Lens    
Spain                           Gwangju World Cup Stadium         Gwangju    
Spain                            Jeonju World Cup Stadium          Jeonju    
Spain                             Suwon World Cup Stadium           Suwon    
Spain                           Gwangju World Cup Stadium         Gwangju    
Spain                          

Como se puede observar en cada dataframe tenemos la indormacion delos aprtidos que ha jugado España tanto local como visitante 

Parte opcional, para correr el siguiente codigo se necesitaba limpiar el kernel debido a que las clumnas son tomadas como indices, por lo cual, preferi poner esta parte opcional antes para que corra todo el codigo sin intervenir

In [30]:
df_partidos = pd.read_csv("WorldCupMatches.csv")
df_partidos[(df_partidos['Home Team Name']=='Spain') | (df_partidos['Away Team Name']=='Spain') ]

Year               Datetime              Stage  \
23   1934   27 May 1934 - 16:30   Preliminary round   
28   1934   31 May 1934 - 16:30      Quarter-finals   
30   1934   01 Jun 1934 - 16:30      Quarter-finals   
55   1950   25 Jun 1950 - 15:00             Group 2   
60   1950   29 Jun 1950 - 15:00             Group 2   
64   1950   02 Jul 1950 - 15:00             Group 2   
69   1950   09 Jul 1950 - 15:00             Group 6   
71   1950   13 Jul 1950 - 15:00             Group 6   
73   1950   16 Jul 1950 - 15:00             Group 6   
141  1962   31 May 1962 - 15:00             Group 3   
149  1962   03 Jun 1962 - 15:00             Group 3   
153  1962   06 Jun 1962 - 15:00             Group 3   
174  1966   13 Jul 1966 - 19:30             Group 2   
177  1966   15 Jul 1966 - 19:30             Group 2   
190  1966   20 Jul 1966 - 19:30             Group 2   
275  1978   03 Jun 1978 - 13:45             Group 3   
282  1978   07 Jun 1978 - 13:45             Group 3   
291  1978   11 Jun 1978 - 13:45             Group 3   
316  1982   16 Jun 1982 - 21:00             Group 5   
328  1982   20 Jun 1982 - 21:00             Group 5   
343  1982   25 Jun 1982 - 21:00             Group 5   
351  1982   02 Jul 1982 - 21:00             Group 2   
355  1982   05 Jul 1982 - 21:00             Group 2   
362  1986   01 Jun 1986 - 12:00             Group D   
378  1986   07 Jun 1986 - 12:00             Group D   
393  1986   12 Jun 1986 - 12:00             Group D   
403  1986   18 Jun 1986 - 16:00         Round of 16   
406  1986   22 Jun 1986 - 16:00      Quarter-finals   
423  1990   13 Jun 1990 - 17:00             Group E   
434  1990   17 Jun 1990 - 21:00             Group E   
445  1990   21 Jun 1990 - 17:00             Group E   
454  1990   26 Jun 1990 - 17:00         Round of 16   
464  1994   17 Jun 1994 - 19:30             Group C   
476  1994   21 Jun 1994 - 16:00             Group C   
491  1994   27 Jun 1994 - 16:00             Group C   
501  1994   02 Jul 1994 - 16:30         Round of 16   
508  1994   09 Jul 1994 - 12:00      Quarter-finals   
523  1998   13 Jun 1998 - 14:30             Group D   
539  1998   19 Jun 1998 - 21:00             Group D   
554  1998   24 Jun 1998 - 21:00             Group D   
587  2002   02 Jun 2002 - 20:30             Group B   
602  2002  07 June 2002 - 18:00             Group B   
619  2002   12 Jun 2002 - 20:30             Group B   
630  2002   16 Jun 2002 - 20:30         Round of 16   
639  2002   22 Jun 2002 - 15:30      Quarter-finals   
658  2006   14 Jun 2006 - 15:00             Group H   
675  2006   19 Jun 2006 - 21:00             Group H   
689  2006   23 Jun 2006 - 16:00             Group H   
699  2006   27 Jun 2006 - 21:00         Round of 16   
723  2010   16 Jun 2010 - 16:00             Group H   
739  2010   21 Jun 2010 - 20:30             Group H   
754  2010   25 Jun 2010 - 20:30             Group H   
763  2010   29 Jun 2010 - 20:30         Round of 16   
767  2010   03 Jul 2010 - 20:30      Quarter-finals   
769  2010   07 Jul 2010 - 20:30         Semi-finals   
771  2010   11 Jul 2010 - 20:30               Final   
774  2014   13 Jun 2014 - 16:00             Group B   
790  2014   18 Jun 2014 - 16:00             Group B   
804  2014   23 Jun 2014 - 13:00             Group B   

                                       Stadium               City  \
23                              Luigi Ferraris             Genoa    
28                              Giovanni Berta          Florence    
30                              Giovanni Berta          Florence    
55                            Durival de Brito          Curitiba    
60   Maracan� - Est�dio Jornalista M�rio Filho    Rio De Janeiro    
64   Maracan� - Est�dio Jornalista M�rio Filho    Rio De Janeiro    
69                                    Pacaembu         Sao Paulo    
71   Maracan� - Est�dio Jornalista M�rio Filho    Rio De Janeiro    
73                                    Pacaembu         Sao Paulo    
141           

## Calcular el vencedor de cada partido

El DataFrame `df_partidos` tiene la información de cada partido, incluyendo los goles marcados por cada equipo, pero no disponemos de una columna que indique si un equipo ganó o perdió. Vamos a crear una columna llamada `Ganador` que indique el nombre del equipo vencedor (con más goles). También podemos encontrar empates, en cuyo caso usaremos `pd.NA` para indicar que desconocemos el vencedor de dicho partido.

**Nota:** Se puede resolver realizando varias asignaciones, o una sola utilizando la función [pd.where](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.where.html).

In [31]:
df_partidos = pd.read_csv("WorldCupMatches.csv")
# alternativa 1
df_partidos["Ganador"] = df_partidos["Home Team Name"]
df_partidos.loc[df_partidos["Home Team Goals"] < df_partidos["Away Team Goals"], "Ganador"] = df_partidos["Away Team Name"]
df_partidos.loc[df_partidos["Home Team Goals"] == df_partidos["Away Team Goals"], "Ganador"] = pd.NA

# alternativa 2
df_partidos["Ganador"] = (
    df_partidos["Home Team Name"].where(
        df_partidos["Home Team Goals"] > df_partidos["Away Team Goals"],
        df_partidos["Away Team Name"].where(df_partidos["Home Team Goals"] < df_partidos["Away Team Goals"], pd.NA)
    )
)

## 11. Número de victorias por país

Calcular para cada país el número de partidos ganados. Para ello, usar la función [dropna](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dropna.html) para eliminar las filas correspondientes a partidos donde la columna `Ganador` es un valor perdido (`na`). Ordenar los países por número de victorias de forma descendente.

Antes de comenzar a a plicar funciones es bueno saber cuantso datos perdidos o na hay en la fila columnas 

In [32]:
df_partidos.isnull().sum()

Year                      0
Datetime                  0
Stage                     0
Stadium                   0
City                      0
Home Team Name            0
Home Team Goals           0
Away Team Goals           0
Away Team Name            0
Win conditions            0
Attendance                2
Half-time Home Goals      0
Half-time Away Goals      0
Referee                   0
Assistant 1               0
Assistant 2               0
RoundID                   0
MatchID                   0
Home Team Initials        0
Away Team Initials        0
Ganador                 190
dtype: int64

Una vez que tengamos los datos pasamos a eliminar las filas que tengan na en la columna ganadores 

In [33]:
df_partidos = df_partidos[df_partidos["Ganador"].notna()]
df_partidos.isnull().sum()

Year                    0
Datetime                0
Stage                   0
Stadium                 0
City                    0
Home Team Name          0
Home Team Goals         0
Away Team Goals         0
Away Team Name          0
Win conditions          0
Attendance              2
Half-time Home Goals    0
Half-time Away Goals    0
Referee                 0
Assistant 1             0
Assistant 2             0
RoundID                 0
MatchID                 0
Home Team Initials      0
Away Team Initials      0
Ganador                 0
dtype: int64

Una vez hecho lo anterior pasamos a contar cuantas veces ha ganado el equipo con la funcion values_counts y especificamos el orden, en este caso descentente 

In [34]:
df_partidos["Ganador"].value_counts(ascending=False)

Brazil                        71
Italy                         45
Argentina                     44
Germany FR                    36
Germany                       34
France                        29
Netherlands                   29
Spain                         29
England                       26
Uruguay                       20
Yugoslavia                    16
Sweden                        16
Soviet Union                  15
Poland                        15
Belgium                       15
Hungary                       15
Mexico                        14
Portugal                      13
Austria                       12
Czechoslovakia                11
Chile                         11
Switzerland                   11
USA                            8
Romania                        8
Denmark                        8
Colombia                       8
Paraguay                       7
Croatia                        7
Costa Rica                     5
Korea Republic                 5
Nigeria   

## Opcional 1

Calcular para cada país el ratio de victorias, calculado como el número de partidos ganados dividido por el número de partidos jugados. Ordenar los países por ratio de forma descendente.

**Pista:** Si obtenemos dos dataframes, uno con las victorias por cada país y otro con el número total de partidos jugados, podemos utilizar la función [pd.merge](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.merge.html) para combinar ambos en un solo DataFrame, y despues crear una nueva columna que sea el ratio.

In [35]:
df_victorias=pd.Series(df_partidos["Ganador"].value_counts())
df_victorias = df_victorias.to_frame()
df_victorias

Ganador
Brazil                           71
Italy                            45
Argentina                        44
Germany FR                       36
Germany                          34
France                           29
Netherlands                      29
Spain                            29
England                          26
Uruguay                          20
Yugoslavia                       16
Sweden                           16
Soviet Union                     15
Poland                           15
Belgium                          15
Hungary                          15
Mexico                           14
Portugal                         13
Austria                          12
Czechoslovakia                   11
Chile                            11
Switzerland                      11
USA                               8
Romania                           8
Denmark                           8
Colombia                          8
Paraguay                          7
Croatia                           7
Costa Rica                        5
Korea Republic                    5
Nigeria                           5
Turkey                            5
Ghana                             4
Japan                             4
Peru                              4
Ecuador                           4
Cameroon                          4
Scotland                          4
Northern Ireland                  3
C�te d'Ivoire                     3
Bulgaria                          3
Algeria                           3
Norway                            2
Russia                            2
Ukraine                           2
South Africa                      2
Morocco                           2
rn">Republic of Ireland           2
Greece                            2
Australia                         2
Saudi Arabia                      2
German DR                         2
Senegal                           2
Slovenia                          1
Czech Republic                    1
Slovakia                          1
Jamaica                           1
rn">Bosnia and Herzegovina        1
Korea DPR                         1
Iran                              1
Wales                             1
Serbia                            1
Cuba                              1
Tunisia                           1

In [36]:
df_victorias.columns

Index(['Ganador'], dtype='object')

In [37]:
df_jugados = pd.Series( df_paises["Country"].value_counts())
df_jugados = df_jugados.to_frame()
df_jugados

Country
Brazil                          108
Italy                            83
Argentina                        81
Germany FR                       62
England                          62
France                           61
Spain                            59
Netherlands                      54
Mexico                           54
Uruguay                          52
Germany                          48
Sweden                           46
Belgium                          43
Yugoslavia                       37
USA                              34
Switzerland                      34
Chile                            34
Hungary                          32
Poland                           31
Korea Republic                   31
Soviet Union                     31
Czechoslovakia                   30
Austria                          29
Paraguay                         27
Bulgaria                         26
Portugal                         26
Scotland                         23
Cameroon                         23
Romania                          21
Colombia                         20
Nigeria                          19
Japan                            17
Costa Rica                       17
Croatia                          16
Denmark                          16
Peru                             15
Algeria                          14
Northern Ireland                 13
Morocco                          13
Saudi Arabia                     13
Australia                        13
rn">Republic of Ireland          13
Ghana                            12
Tunisia                          12
Greece                           11
Ecuador                          10
Turkey                           10
C�te d'Ivoire                     9
Honduras                          9
Russia                            9
South Africa                      9
Norway                            8
Korea DPR                         7
El Salvador                       6
German DR                         6
IR Iran                           6
New Zealand                       6
Iran                              6
Slovenia                          6
Bolivia                           6
Wales                             5
Ukraine                           5
Senegal                           5
Slovakia                          4
Egypt                             4
rn">Serbia and Montenegro         3
China PR                          3
Jamaica                           3
Togo                              3
Canada                            3
Kuwait                            3
Israel                            3
Serbia                            3
rn">United Arab Emirates          3
rn">Trinidad and Tobago           3
Iraq                              3
Zaire                             3
Cuba                              3
Haiti                             3
Angola                            3
rn">Bosnia and Herzegovina        3
Czech Republic                    3
Dutch East Indies                 1

In [38]:
df_ratio=df_jugados.join(df_victorias,rsuffix=" ")
df_ratio

Country  Ganador
Brazil                          108    71.00
Italy                            83    45.00
Argentina                        81    44.00
Germany FR                       62    36.00
England                          62    26.00
France                           61    29.00
Spain                            59    29.00
Netherlands                      54    29.00
Mexico                           54    14.00
Uruguay                          52    20.00
Germany                          48    34.00
Sweden                           46    16.00
Belgium                          43    15.00
Yugoslavia                       37    16.00
USA                              34     8.00
Switzerland                      34    11.00
Chile                            34    11.00
Hungary                          32    15.00
Poland                           31    15.00
Korea Republic                   31     5.00
Soviet Union                     31    15.00
Czechoslovakia                   30    11.00
Austria                          29    12.00
Paraguay                         27     7.00
Bulgaria                         26     3.00
Portugal                         26    13.00
Scotland                         23     4.00
Cameroon                         23     4.00
Romania                          21     8.00
Colombia                         20     8.00
Nigeria                          19     5.00
Japan                            17     4.00
Costa Rica                       17     5.00
Croatia                          16     7.00
Denmark                          16     8.00
Peru                             15     4.00
Algeria                          14     3.00
Northern Ireland                 13     3.00
Morocco                          13     2.00
Saudi Arabia                     13     2.00
Australia                        13     2.00
rn">Republic of Ireland          13     2.00
Ghana                            12     4.00
Tunisia                          12     1.00
Greece                           11     2.00
Ecuador                          10     4.00
Turkey                           10     5.00
C�te d'Ivoire                     9     3.00
Honduras                          9      nan
Russia                            9     2.00
South Africa                      9     2.00
Norway                            8     2.00
Korea DPR                         7     1.00
El Salvador                       6      nan
German DR                         6     2.00
IR Iran                           6      nan
New Zealand                       6      nan
Iran                              6     1.00
Slovenia                          6     1.00
Bolivia                           6      nan
Wales                             5     1.00
Ukraine                           5     2.00
Senegal                           5     2.00
Slovakia                          4     1.00
Egypt                             4      nan
rn">Serbia and Montenegro         3      nan
China PR                          3      nan
Jamaica                           3     1.00
Togo                              3      nan
Canada                            3      nan
Kuwait                            3      nan
Israel                            3      nan
Serbia                            3     1.00
rn">United Arab Emirates          3      nan
rn">Trinidad and Tobago           3      nan
Iraq                              3      nan
Zaire                             3      nan
Cuba                              3     1.00
Haiti                             3      nan
Angola                            3      nan
rn">Bosnia and Herzegovina        3     1.00
Czech Republic                    3     1.00
Dutch East Indies                 1      nan

In [40]:
df_ratio.assign(Ratio=df_ratio["Ganador"]/df_ratio["Country"])

Country  Ganador  Ratio
Brazil                          108    71.00   0.66
Italy                            83    45.00   0.54
Argentina                        81    44.00   0.54
Germany FR                       62    36.00   0.58
England                          62    26.00   0.42
France                           61    29.00   0.48
Spain                            59    29.00   0.49
Netherlands                      54    29.00   0.54
Mexico                           54    14.00   0.26
Uruguay                          52    20.00   0.38
Germany                          48    34.00   0.71
Sweden                           46    16.00   0.35
Belgium                          43    15.00   0.35
Yugoslavia                       37    16.00   0.43
USA                              34     8.00   0.24
Switzerland                      34    11.00   0.32
Chile                            34    11.00   0.32
Hungary                          32    15.00   0.47
Poland                           31    15.00   0.48
Korea Republic                   31     5.00   0.16
Soviet Union                     31    15.00   0.48
Czechoslovakia                   30    11.00   0.37
Austria                          29    12.00   0.41
Paraguay                         27     7.00   0.26
Bulgaria                         26     3.00   0.12
Portugal                         26    13.00   0.50
Scotland                         23     4.00   0.17
Cameroon                         23     4.00   0.17
Romania                          21     8.00   0.38
Colombia                         20     8.00   0.40
Nigeria                          19     5.00   0.26
Japan                            17     4.00   0.24
Costa Rica                       17     5.00   0.29
Croatia                          16     7.00   0.44
Denmark                          16     8.00   0.50
Peru                             15     4.00   0.27
Algeria                          14     3.00   0.21
Northern Ireland                 13     3.00   0.23
Morocco                          13     2.00   0.15
Saudi Arabia                     13     2.00   0.15
Australia                        13     2.00   0.15
rn">Republic of Ireland          13     2.00   0.15
Ghana                            12     4.00   0.33
Tunisia                          12     1.00   0.08
Greece                           11     2.00   0.18
Ecuador                          10     4.00   0.40
Turkey                           10     5.00   0.50
C�te d'Ivoire                     9     3.00   0.33
Honduras                          9      nan    nan
Russia                            9     2.00   0.22
South Africa                      9     2.00   0.22
Norway                            8     2.00   0.25
Korea DPR                         7     1.00   0.14
El Salvador                       6      nan    nan
German DR                         6     2.00   0.33
IR Iran                           6      nan    nan
New Zealand                       6      nan    nan
Iran                              6     1.00   0.17
Slovenia                          6     1.00   0.17
Bolivia                           6      nan    nan
Wales                             5     1.00   0.20
Ukraine                           5     2.00   0.40
Senegal                           5     2.00   0.40
Slovakia                          4     1.00   0.25
Egypt                             4      nan    nan
rn">Serbia and Montenegro         3      nan    nan
China PR                          3      nan    nan
Jamaica                           3     1.00   0.33
Togo                              3      nan    nan
Canada                            3      nan    nan
Kuwait                            3      nan    nan
Israel                            3      nan    nan
Serbia                            3     1.00   0.33
rn">United Arab Emirates          3      nan    nan
rn">Trinidad and Tobago           3      nan    nan
Iraq                              3      nan    nan
Zaire                   

Para sacar el ratio se tuco que crear dos 3 dataframes, dos con la infomacion que recopilamos de sus antecesores y uno donde se agrego la columna de ratio con sus respectivos valores, es decir con la operacion necesaria

## Opcional 2

¿Se te ocurre alguna pregunta adicional que hacerle a los datos? Utiliza una celda de tipo markdown para documentarla y otra de código para implementarla y ver así la respuesta.